internal/handlers/user.go

```diff
@@ -53,50 +55,53 @@ func (api *API) createUserIfNotExists(ctx context.Context, id string, userName s
 	var uuid uuid.UUID
 
 	err := api.transaction(ctx, func(tx *gorm.DB) error {
-		// First lets check if the user has ever existed in the database
-		res := tx.Unscoped().First(&user, "id = ?", id)
-
-		// If the user exists, then lets restore their status in the database
-		if res.Error == nil {
-			if user.DeletedAt.Valid {
-				user.DeletedAt = gorm.DeletedAt{}
-				if res := tx.Unscoped().Model(&user).Update("DeletedAt", user.DeletedAt); res.Error != nil {
-					return res.Error
+		// Retry the operation if we get a duplicate key error which can occur on concurrent requests when creating a user
+		return util.RetryOperationForErrors(ctx, time.Millisecond*10, 1, []error{gorm.ErrDuplicatedKey}, func() error {
```

Comment:
Perhaps we should do this outside the transaction block? I'm slightly worried that you might commit a change that occurred in an operation that had failed.

Reply:
Good catch, definitely right, will fix this up!